x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java

```diff
@@ -0,0 +1,434 @@
+import static org.elasticsearch.xpack.core.security.authc.oidc.OpenIdConnectRealmSettings.HTTP_SOCKET_TIMEOUT;
+
+/**
+ * Handles an OpenID Connect Authentication response as received by the facilitator. In the case of an implicit flow, validates
+ * the ID Token and extracts the elasticsearch user properties from it. In the case of an authorization code flow, it first
+ * exchanges the code in the authentication response for an ID Token at the token endpoint of the OpenID Connect Provider.
+ */
+public class OpenIdConnectAuthenticator {
+
+    private final RealmConfig realmConfig;
+    private final OpenIdConnectProviderConfiguration opConfig;
+    private final RelyingPartyConfiguration rpConfig;
+    private final SSLService sslService;
+    private final PrivilegedResourceRetriever privilegedResourceResolver;
+
+    protected final Logger logger = LogManager.getLogger(getClass());
+
+    public OpenIdConnectAuthenticator(RealmConfig realmConfig, OpenIdConnectProviderConfiguration opConfig,
+                                      RelyingPartyConfiguration rpConfig,
+                                      SSLService sslService) {
+        this.realmConfig = realmConfig;
+        this.opConfig = opConfig;
+        this.rpConfig = rpConfig;
+        this.sslService = sslService;
+        this.privilegedResourceResolver = getPrivilegedResourceRetriever();
+    }
+
+    // For testing
+    OpenIdConnectAuthenticator(RealmConfig realmConfig, OpenIdConnectProviderConfiguration opConfig, RelyingPartyConfiguration rpConfig,
+                               SSLService sslService, PrivilegedResourceRetriever resourceRetriever) {
+        this.realmConfig = realmConfig;
+        this.opConfig = opConfig;
+        this.rpConfig = rpConfig;
+        this.sslService = sslService;
+        this.privilegedResourceResolver = resourceRetriever;
+    }
+
+    OpenIdConnectAuthenticator() {
+        this(null, null, null, null, null);
+    }
+
+    /**
+     * Processes an OpenID Connect Response to an Authentication Request that comes in the form of a URL with the necessary parameters, that
+     * is contained in the provided Token. If the response is valid, it returns a set of OpenID Connect claims that identify the
+     * authenticated user. If the UserInfo endpoint is specified in the configuration, we attempt to make a UserInfo request and add
+     * the returned claims.
+     *
+     * @param token The OpenIdConnectToken to consume
+     * @return a {@link JWTClaimsSet} with the OP claims for the user
+     * @throws ElasticsearchSecurityException if the response is invalid in any way
+     */
+    public JWTClaimsSet authenticate(OpenIdConnectToken token) {
+        try {
+            AuthenticationResponse authenticationResponse = AuthenticationResponseParser.parse(new URI(token.getRedirectUrl()));
+            Nonce expectedNonce = new Nonce(token.getNonce());
+            State expectedState = new State(token.getState());
```

Comment:
I think it would be neater if `token.getNonce()` would return a `Nonce` and `token.getState()` would return a `State` .